File: packages/client-android/src/main/kotlin/dev/poppy/android/Validate.kt
/*
 * Validation entry point for the Android renderer. Per ADR-0009, decoding IS
 * the validator: we run the typed kotlinx.serialization decoder and map any
 * SerializationException to a [ValidationError] with the corpus keyword
 * declared in the seven invalid cases under
 * `packages/conformance/cases/invalid/`.
 *
 * Mapping table (ADR-0009, repeated here for code locality):
 *
 *   | Exception                                | Keyword       |
 *   |------------------------------------------|---------------|
 *   | Unknown / missing class discriminator    | discriminator |
 *   | MissingFieldException                    | required      |
 *   | JsonDecodingException (type mismatch)    | type          |
 *   | Custom version-compat check              | version       |
 *
 * Order of checks mirrors `@poppy/server-ts/src/validate.ts`:
 *   1. Decode (catches discriminator/required/type errors).
 *   2. Version-compat check (catches version errors). Only runs if decode passed.
 */

package dev.poppy.android

import kotlinx.serialization.MissingFieldException
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json

/** Wire-format version this renderer targets. See ADR-0006. */
const val SCHEMA_VERSION: String = "0.1"

/**
 * Structural error in a Poppy document. Mirrors `@poppy/server-ts`'s
 * `ValidationError` shape.
 */
data class ValidationError(
    /** JSON pointer to the location of the failure, e.g. `/root/children/0/value`. */
    val path: String,
    /** Schema keyword that failed (`discriminator`, `required`, `type`, `version`). */
    val keyword: String,
    /** Human-readable message from the underlying decoder. */
    val message: String,
)

/** Result of parsing + validating a Poppy document. Never throws. */
sealed class ValidationResult {
    data class Ok(val document: PoppyDocument) : ValidationResult()
    data class Failure(val errors: List<ValidationError>) : ValidationResult()
}

/**
 * The Poppy public API entry. Mirrors the contract of `@poppy/server-ts`'s
 * [validate]/[isValid] functions.
 */
object Poppy {
    /**
     * Configured JSON instance. `classDiscriminator = "type"` matches the
     * wire-format discriminator key. `ignoreUnknownKeys = true` honors
     * ADR-0006's forward-compat rule: a v0.2 document with a new optional
     * field decodes cleanly on a v0.1 renderer (the field is dropped).
     */
    internal val json: Json = Json {
        classDiscriminator = "type"
        ignoreUnknownKeys = true
    }

    /** Parse and validate a document. Never throws on malformed input. */
    fun validate(jsonString: String): ValidationResult {
        val decoded: PoppyDocument = try {
            json.decodeFromString(PoppyDocument.serializer(), jsonString)
        } catch (e: MissingFieldException) {
            return ValidationResult.Failure(listOf(mapMissingField(e)))
        } catch (e: SerializationException) {
            return ValidationResult.Failure(listOf(mapSerializationError(e)))
        } catch (e: IllegalArgumentException) {
            // kotlinx.serialization throws this for some malformed input shapes.
            return ValidationResult.Failure(listOf(mapSerializationError(e)))
        }

        val versionError = checkVersionCompat(decoded.version)
        if (versionError != null) {
            return ValidationResult.Failure(listOf(versionError))
        }
        return ValidationResult.Ok(decoded)
    }

    /** Type-guard form of [validate]. */
    fun isValid(jsonString: String): Boolean = validate(jsonString) is ValidationResult.Ok
}

// --- Internal: error mapping -----------------------------------------------

private val SUPPORTED_VERSION: Pair<Int, Int> = parseVersion(SCHEMA_VERSION)
private val SUPPORTED_MAJOR: Int get() = SUPPORTED_VERSION.first
private val SUPPORTED_MINOR: Int get() = SUPPORTED_VERSION.second

private fun parseVersion(v: String): Pair<Int, Int> {
    val parts = v.split(".")
    require(parts.size == 2) { "version must be MAJOR.MINOR; got '$v'" }
    return parts[0].toInt() to parts[1].toInt()
}

private fun checkVersionCompat(version: String): ValidationError? {
    val (major, minor) = try {
        parseVersion(version)
    } catch (_: Exception) {
        // Decoding succeeded, so `version` is a string; but it might not
        // match MAJOR.MINOR. The schema enforces the pattern; mobile decoder
        // doesn't, so we defend here. Treat as a version error.
        return ValidationError(
            path = "/version",
            keyword = "version",
            message = "malformed document version \"$version\"; expected MAJOR.MINOR",
        )
    }
    if (major != SUPPORTED_MAJOR) {
        return ValidationError(
            path = "/version",
            keyword = "version",
            message = "unsupported document major version \"$version\"; this renderer supports $SUPPORTED_MAJOR.x",
        )
    }
    if (minor > SUPPORTED_MINOR) {
        return ValidationError(
            path = "/version",
            keyword = "version",
            message = "document minor version $minor exceeds renderer's supported minor $SUPPORTED_MINOR",
        )
    }
    return null
}

private fun mapMissingField(e: MissingFieldException): ValidationError {
    // MissingFieldException doesn't expose a JSON pointer; the message is the
    // best signal we have. The corpus's required-keyword cases (003, 005) all
    // target a missing field on the root component, so reporting `/root` is
    // accurate for the corpus and a sensible default for nested cases.
    val fields = extractMissingFields(e.message ?: "")
    val msg = if (fields.isNotEmpty()) {
        "missing required field(s): ${fields.joinToString(", ")}"
    } else {
        e.message ?: "missing required field"
    }
    return ValidationError(path = "/root", keyword = "required", message = msg)
}

private fun extractMissingFields(message: String): List<String> {
    // kotlinx.serialization's MissingFieldException message format is:
    //   "Field 'foo' is required for type with serial name '...', but it was missing"
    //   "Fields [a, b] are required for type with serial name '...', but they were missing"
    val singleField = Regex("Field '([^']+)' is required").find(message)?.groupValues?.get(1)
    if (singleField != null) return listOf(singleField)
    val multiField = Regex("Fields \\[([^\\]]+)\\] are required").find(message)?.groupValues?.get(1)
    if (multiField != null) return multiField.split(",").map { it.trim() }
    return emptyList()
}

private fun mapSerializationError(e: Throwable): ValidationError {
    val message = e.message ?: "decoding failed"

    // Discriminator failures come in a few shapes depending on whether the
    // discriminator field is missing or has an unknown value. kotlinx.serialization
    // surfaces both as SerializationException with a recognizable message.
    if (looksLikeDiscriminatorError(message)) {
        return ValidationError(
            path = pathFromMessage(message, default = "/root"),
            keyword = "discriminator",
            message = message,
        )
    }
    // Type mismatch: e.g. Text.value being an Int. kotlinx surfaces this as
    // "Unexpected JSON token" or includes the expected primitive name.
    if (looksLikeTypeError(message)) {
        return ValidationError(
            path = pathFromMessage(message, default = "/root"),
            keyword = "type",
            message = message,
        )
    }
    // Best-effort fallback. The corpus only relies on
    // discriminator/required/type/version; anything that falls through here is
    // necessarily beyond the corpus's coverage.
    return ValidationError(path = "/", keyword = "type", message = message)
}

private fun looksLikeDiscriminatorError(message: String): Boolean {
    return message.contains("Polymorphic serializer was not found") ||
        message.contains("class discriminator") ||
        message.contains("Class discriminator") ||
        message.contains("Serializer for subclass") ||
        message.contains("is not registered") ||
        message.contains("type") && message.contains("not found")
}

private fun looksLikeTypeError(message: String): Boolean {
    return message.contains("Unexpected JSON token") ||
        message.contains("but had") ||
        message.contains("Failed to parse") ||
        message.contains("Expected ") ||
        message.contains("kotlin.String") ||
        message.contains("kotlin.Int")
}

private fun pathFromMessage(message: String, default: String): String {
    // kotlinx.serialization sometimes embeds a JSON path in the message after
    // "at path: ". Extract it if present so test output stays informative.
    val match = Regex("at path:?\\s*(\\$[^\\s]+|/[^\\s]+)").find(message)
    val raw = match?.groupValues?.get(1) ?: return default
    // Normalize JSONPath ($.root.value) to JSON Pointer (/root/value) for parity
    // with @poppy/server-ts's error shape.
    return when {
        raw.startsWith("$.") -> "/" + raw.removePrefix("$.").replace(".", "/")
        raw.startsWith("$") -> "/" + raw.removePrefix("$").replace(".", "/")
        else -> raw
    }
}

File: packages/client-ios/Sources/Poppy/Validate.swift
// Validate Poppy v0.1 documents.
//
// Per ADR-0009, decoding *is* the validator on iOS. The typed Codable tree
// (see Document.swift) catches missing required fields, wrong-typed fields,
// and missing/unknown component discriminators natively. After a successful
// decode, a hand-written version-compat check (per ADR-0006) rejects
// documents whose major doesn't match SUPPORTED_MAJOR or whose minor exceeds
// SUPPORTED_MINOR. Together these cover every keyword in the corpus.
//
// Maps DecodingError → ValidationError keyword:
//
//   .keyNotFound          → "required"
//   .typeMismatch         → "type"
//   .dataCorrupted        → "discriminator" (our custom Component/Action throw it
//                           for missing/unknown `type`)
//   .valueNotFound        → "required" (treats null on a non-optional field
//                           as effectively a missing field)
//   anything else          → "type" (a defensive fallback; all kinds we expect
//                           to see in the corpus are explicitly handled above)
//
// The path is built from the DecodingError's coding-key chain, prefixed with
// "/" — matching the JSON-pointer style @poppy/server-ts emits.

import Foundation

/// Wire-format version this build of Poppy supports. Mirrors `SCHEMA_VERSION`
/// in packages/schema/src/types.ts.
public let SCHEMA_VERSION: String = "0.1"

public struct ValidationError: Error, Sendable, Equatable {
    /// JSON pointer to the location of the failure, e.g. "/root/children/0/value".
    public let path: String
    /// The conformance keyword that failed: "discriminator", "required", "type", "version".
    public let keyword: String
    /// Human-readable message from the underlying decoder, lightly cleaned.
    public let message: String

    public init(path: String, keyword: String, message: String) {
        self.path = path
        self.keyword = keyword
        self.message = message
    }
}

public enum ValidationResult: Sendable {
    case ok(PoppyDocument)
    case failure([ValidationError])

    public var isOk: Bool {
        if case .ok = self { return true }
        return false
    }

    public var document: PoppyDocument? {
        if case .ok(let d) = self { return d }
        return nil
    }

    public var errors: [ValidationError] {
        if case .failure(let e) = self { return e }
        return []
    }
}

/// Top-level entry point for the renderer's validation. Mirrors the public
/// API in @poppy/server-ts. Never throws on user input.
///
/// Order of checks:
///   1. Decode JSON → typed tree. On failure, map the DecodingError to a
///      ValidationError with the appropriate keyword and return Failure.
///   2. If decoding succeeded, run the version-compat check. If incompatible,
///      return Failure with a single version error.
///   3. Otherwise return Ok(document).
public enum Poppy {
    public static func validate(_ json: Data) -> ValidationResult {
        let decoder = JSONDecoder()
        let document: PoppyDocument
        do {
            document = try decoder.decode(PoppyDocument.self, from: json)
        } catch let decodingError as DecodingError {
            return .failure([mapDecodingError(decodingError)])
        } catch {
            return .failure([
                ValidationError(
                    path: "/",
                    keyword: "type",
                    message: "JSON parse failure: \(error.localizedDescription)"
                ),
            ])
        }

        if let versionError = checkVersionCompat(document.version) {
            return .failure([versionError])
        }

        return .ok(document)
    }

    /// Type-guard form. Returns true iff `validate(json)` is `.ok`.
    public static func isValid(_ json: Data) -> Bool {
        return validate(json).isOk
    }
}

// MARK: - Internals

/// Map a DecodingError to a ValidationError. Public for unit tests; not
/// part of the renderer's stable API.
public func mapDecodingError(_ error: DecodingError) -> ValidationError {
    switch error {
    case .keyNotFound(let key, let context):
        let path = jsonPointer(from: context.codingPath)
        return ValidationError(
            path: path,
            keyword: "required",
            message: "missing required field \"\(key.stringValue)\" at \(path.isEmpty ? "/" : path)"
        )
    case .typeMismatch(_, let context):
        let path = jsonPointer(from: context.codingPath)
        return ValidationError(
            path: path.isEmpty ? "/" : path,
            keyword: "type",
            message: context.debugDescription
        )
    case .dataCorrupted(let context):
        // Our custom Component/Action init throws .dataCorrupted for missing
        // or unknown `type`. The codingPath ends at the `type` key in that
        // case, so we trim it back to the parent component object's path —
        // which is what the corpus expected-error.json files reference.
        let path = jsonPointer(from: trimmingTrailingTypeKey(context.codingPath))
        return ValidationError(
            path: path.isEmpty ? "/" : path,
            keyword: "discriminator",
            message: context.debugDescription
        )
    case .valueNotFound(_, let context):
        let path = jsonPointer(from: context.codingPath)
        return ValidationError(
            path: path.isEmpty ? "/" : path,
            keyword: "required",
            message: context.debugDescription
        )
    @unknown default:
        return ValidationError(
            path: "/",
            keyword: "type",
            message: "unexpected DecodingError: \(error)"
        )
    }
}

/// Run the version-compat check. Returns nil on compatible, a populated
/// ValidationError on incompatible. Mirrors @poppy/server-ts's checkVersionCompat.
public func checkVersionCompat(_ version: String) -> ValidationError? {
    let parts = version.split(separator: ".")
    guard parts.count == 2,
          let major = Int(parts[0]),
          let minor = Int(parts[1]) else {
        // The Codable layer doesn't enforce the MAJOR.MINOR pattern (the JSON
        // Schema does), so we get here for malformed strings like "v1" or
        // "1.0.0". Treat as a version error (the only kind that touches
        // /version in the corpus).
        return ValidationError(
            path: "/version",
            keyword: "version",
            message: "malformed document version \"\(version)\"; expected MAJOR.MINOR"
        )
    }

    let supported = parseSupportedVersion()
    if major != supported.major {
        return ValidationError(
            path: "/version",
            keyword: "version",
            message: "unsupported document major version \"\(version)\"; this renderer supports \(supported.major).x"
        )
    }
    if minor > supported.minor {
        return ValidationError(
            path: "/version",
            keyword: "version",
            message: "document minor version \(minor) exceeds renderer's supported minor \(supported.minor)"
        )
    }
    return nil
}

private func parseSupportedVersion() -> (major: Int, minor: Int) {
    let parts = SCHEMA_VERSION.split(separator: ".")
    let major = parts.count > 0 ? Int(parts[0]) ?? 0 : 0
    let minor = parts.count > 1 ? Int(parts[1]) ?? 0 : 0
    return (major, minor)
}

/// Build a JSON pointer from a chain of CodingKey. Numeric coding keys (array
/// indexes) are kept as their integer; everything else as the stringValue.
/// Empty input yields the empty string (the caller substitutes "/" if needed).
private func jsonPointer(from codingPath: [CodingKey]) -> String {
    if codingPath.isEmpty { return "" }
    var parts: [String] = []
    for key in codingPath {
        if let intValue = key.intValue {
            parts.append("\(intValue)")
        } else {
            parts.append(escapePointerToken(key.stringValue))
        }
    }
    return "/" + parts.joined(separator: "/")
}

/// Escape `~` and `/` per RFC 6901.
private func escapePointerToken(_ token: String) -> String {
    return token
        .replacingOccurrences(of: "~", with: "~0")
        .replacingOccurrences(of: "/", with: "~1")
}

/// Drop a trailing `type` key from a coding-path. Used for discriminator
/// errors so the path points to the component object, not its `type` field —
/// matching the corpus convention.
private func trimmingTrailingTypeKey(_ codingPath: [CodingKey]) -> [CodingKey] {
    if let last = codingPath.last, last.stringValue == "type" {
        return Array(codingPath.dropLast())
    }
    return codingPath
}
